test/tools/ossfuzz/protoToAbiV2.cpp

```diff
@@ -810,13 +810,15 @@ pair<string, string> AssignCheckVisitor::visit(ArrayType const& _type)
 		length = getDynArrayLengthFromFuzz(_type.length(), counter());
 		lengthStr = to_string(length);
 		if (m_stateVar)
-			resizeBuffer = assignAndCheckStringPair(
-				m_varName + ".length",
-				m_paramName + ".length",
-				lengthStr,
-				lengthStr,
-				DataType::VALUE
-				);
+		{
+			// Dynamic storage arrays are resized via the empty push() operation
+			resizeBuffer.first = Whiskers(R"(for(uint i = 0; i < <length>; i++) <arrayRef>.push();)")
+				("length", lengthStr)
+				("arrayRef", m_varName)
+				.render();
+			// Add a dynamic check on the resized length
+			resizeBuffer.second = checkString(m_paramName + ".length", lengthStr, DataType::VALUE);
```

Comment:
This adds a check on the resized array. Example ```if (x.length != 2) return -1;```

Reply:
Is this correct for the general case?
In the code before you increase the length by `lengthStr`, but if the length was not 0 before, this will fail.

Reply:
The length before this code is executed must be zero always because this code path is taken for dynamically sized storage arrays only.

```
// Implicit length of dynamic array is zero
bool[] x;
for (uint i = 0; i < 4; i++) x.push();
```

so that we can assume length equals the number of push operations executed for that dimension.

Reply:
Well, but what if there are two loops doing that for the same state variable?

Reply:
It does not matter. We can have at most one loop per array dimension. All length checks are specific to a given dimension, so even if we have multiple loops per state variable, each of those loops are resizing starting from `0` for that dimension.

I think what you are thinking about is what if there is more than one loop per array reference (e.g., `x[0]`, `x[0][1]` etc.), yes that's going to be a problem. However, this PR creates at most one loop per array reference

Here's what I mean

```
contract C {
   bool[][] x;
   function foo() public {
      // First loop: Add four outer slots -> x.length = 4
      for (uint i  = 0; i < 4; i++) x.push();
      // Second loop: Add 2-D array in first outer slot -> x[0].length = 2
      for (uint i  = 0; i < 2; i++) x[0].push();     
      ...
   }

   function coder() public {
     // Check the length of the outer array
     if (x.length != 4) return 1;
     // Note that the check is on x[0] and not x
     if (x[0].length != 2) return 2;
     ...
   }
}
```

Reply:
Right, that looks good.
What I was wondering if it could happen was something like
```
contract C {
  uint[] x;
  function f() public {
    for (uint i = 0; i < 4; ++i) x.push();
    // here x.length is 4
    for (uint i = 0; i < 4; ++i) x.push();
    // here x.length is 8
  }
}

Reply:
I'm fairly confident more than one for-loop will never be created per array reference as in your example above. The reason is as follows:

- `m_varName` (the class member of type string whose value is the array reference on which `.push()` is called) is different for each of the array's dimensions.

The change of `m_varName` happens in this for-loop

https://github.com/ethereum/solidity/blob/49aec64cafd06bde9b0f3c46517bee3778531115/test/tools/ossfuzz/protoToAbiV2.cpp#L851-L860

I thought of adding an assertion to encode this but couldn't find a nice way to.

Reply:
Ok sounds good